File: src/levels/L04-singleton/ts/singleton.ts
class TicketCounter {
  private static _instance?: TicketCounter

  // Una sola instancia para toda la cafetería.
  static get instance() { return (this._instance ??= new TicketCounter()) }
  private constructor() {}

  private last = 0

  // region: TicketCounter#next
  next() { return ++this.last }
  // endregion
}

class Register {
  constructor(private display: Display) {}

  // Resuelve el problema… a costa de que cualquier clase pueda llegar al contador global.
  // region: Register#checkout
  checkout(order: Order) {
    order.ticket = TicketCounter.instance.next()
    this.display.show(order)
  }
  // endregion
}

File: src/levels/L04-singleton/singleton.rb
class TicketCounter
  # Una sola instancia para toda la cafetería.
  def self.instance = @instance ||= new
  private_class_method :new

  def initialize = @last = 0

  # region: TicketCounter#next
  def next = @last += 1
  # endregion
end

class Register
  def initialize(display) = @display = display

  # Resuelve el problema… a costa de que cualquier clase pueda llegar al contador global.
  # region: Register#checkout
  def checkout(order)
    order.ticket = TicketCounter.instance.next
    @display.show(order)
  end
  # endregion
end
